src/server/services/wmts/qgswmtsgetcapabilities.cpp

```diff
@@ -0,0 +1,759 @@
+    }
+
+    //End
+    return contentsElement;
+  }
+  namespace
+  {
+    QList< layerDef > getWmtsLayerList( QgsServerInterface *serverIface, const QgsProject *project )
+    {
+      QList< layerDef > wmtsLayers;
+#ifdef HAVE_SERVER_PYTHON_PLUGINS
+      QgsAccessControl *accessControl = serverIface->accessControls();
+#endif
+      QgsCoordinateReferenceSystem wgs84 = QgsCoordinateReferenceSystem::fromOgcWmsCrs( GEO_EPSG_CRS_AUTHID );
+
+      QStringList nonIdentifiableLayers = project->nonIdentifiableLayers();
+
+      // WMTS Project configuration
+      bool wmtsProject = project->readBoolEntry( QStringLiteral( "WMTSLayers" ), QStringLiteral( "Project" ) );
+
+      // Root Layer name
+      QString rootLayerName = QgsServerProjectUtils::wmsRootName( *project );
+      if ( rootLayerName.isEmpty() && !project->title().isEmpty() )
+      {
+        rootLayerName = project->title();
+      }
+
+      if ( wmtsProject && !rootLayerName.isEmpty() )
+      {
+        layerDef pLayer;
+        pLayer.id = rootLayerName;
+
+        if ( !project->title().isEmpty() )
+        {
+          pLayer.title = project->title();
+          pLayer.abstract = project->title();
+        }
+
+        //transform the project native CRS into WGS84
+        QgsRectangle projRect = QgsServerProjectUtils::wmsExtent( *project );
+        QgsCoordinateReferenceSystem projCrs = project->crs();
+        QgsCoordinateTransform exGeoTransform( projCrs, wgs84, project );
+        try
+        {
+          pLayer.wgs84BoundingRect = exGeoTransform.transformBoundingBox( projRect );
+        }
+        catch ( const QgsCsException & )
+        {
+          pLayer.wgs84BoundingRect = QgsRectangle( -180, -90, 180, 90 );
+        }
+
+        // Formats
+        bool wmtsPngProject = project->readBoolEntry( QStringLiteral( "WMTSPngLayers" ), QStringLiteral( "Project" ) );
+        if ( wmtsPngProject )
+          pLayer.formats << QStringLiteral( "image/png" );
+        bool wmtsJpegProject = project->readBoolEntry( QStringLiteral( "WMTSJpegLayers" ), QStringLiteral( "Project" ) );
+        if ( wmtsJpegProject )
+          pLayer.formats << QStringLiteral( "image/jpeg" );
+
+        // Project is not queryable in WMS
+        //pLayer.queryable = ( nonIdentifiableLayers.count() != project->count() );
+        pLayer.queryable = false;
+
+        wmtsLayers.append( pLayer );
+      }
+
+      QStringList wmtsGroupNameList = project->readListEntry( QStringLiteral( "WMTSLayers" ), QStringLiteral( "Group" ) );
+      if ( !wmtsGroupNameList.isEmpty() )
+      {
+        QgsLayerTreeGroup *treeRoot = project->layerTreeRoot();
+
+        QStringList wmtsPngGroupNameList = project->readListEntry( QStringLiteral( "WMTSPngLayers" ), QStringLiteral( "Group" ) );
+        QStringList wmtsJpegGroupNameList = project->readListEntry( QStringLiteral( "WMTSJpegLayers" ), QStringLiteral( "Group" ) );
+
+        for ( QString gName : wmtsGroupNameList )
```

Comment:
I think that `gName` should be a const reference in this case.

Reply:
yes